adapter/service/battery/service.go

```diff
@@ -0,0 +1,272 @@
+	// Depending on a caching and reporting configuration the service might decide to skip a report.
+	// To make sure report is being sent regardless of circumstances set the force argument to true.
+	SendBatterySensorReport(force bool) (bool, error)
+	// SendBatteryFullReport sends a full battery report. Returns true if a report was sent.
+	// Depending on a caching and reporting configuration the service might decide to skip a report.
+	// To make sure report is being sent regardless of circumstances set the force argument to true.
+	SendBatteryFullReport(force bool) (bool, error)
+}
+
+// Config represents a service configuration.
+type Config struct {
+	Specification     *fimptype.Service
+	Reporter          Reporter
+	ReportingStrategy cache.ReportingStrategy
+}
+
+// NewService creates a new instance of a battery FIMP service.
+func NewService(
+	mqtt *fimpgo.MqttTransport,
+	cfg *Config,
+) Service {
+	cfg.Specification.Name = Battery
+
+	cfg.Specification.EnsureInterfaces(requiredInterfaces()...)
+
+	if cfg.ReportingStrategy == nil {
+		cfg.ReportingStrategy = DefaultReportingStrategy
+	}
+
+	s := &service{
+		Service:           adapter.NewService(mqtt, cfg.Specification),
+		reporter:          cfg.Reporter,
+		lock:              &sync.Mutex{},
+		reportingCache:    cache.NewReportingCache(),
+		reportingStrategy: cfg.ReportingStrategy,
+	}
+
+	return s
+}
+
+// service is a private implementation of a battery FIMP service.
+type service struct {
+	adapter.Service
+
+	reporter          Reporter
+	lock              *sync.Mutex
+	reportingCache    cache.ReportingCache
+	reportingStrategy cache.ReportingStrategy
+}
+
+// SendBatteryLevelReport sends a battery level report. Returns true if a report was sent.
+// Depending on a caching and reporting configuration the service might decide to skip a report.
+// To make sure report is being sent regardless of circumstances set the force argument to true.
+func (s *service) SendBatteryLevelReport(force bool) (bool, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	level, err := s.reporter.BatteryLevelReport()
+	if err != nil {
+		return false, fmt.Errorf("failed to get battery level: %w", err)
+	}
+
+	if !force && !s.reportingCache.ReportRequired(s.reportingStrategy, EvtLevelReport, "", level) {
+		return false, nil
+	}
+
+	message := fimpgo.NewIntMessage(
+		EvtLevelReport,
+		s.Name(),
+		level,
+		nil,
+		nil,
+		nil,
+	)
+
+	err = s.SendMessage(message)
+	if err != nil {
+		return false, fmt.Errorf("failed to send battery level report: %w", err)
+	}
+
+	s.reportingCache.Reported(EvtLevelReport, "", level)
+
+	return true, nil
+}
+
+// SendBatteryAlarmReport sends a battery alarm report. Returns true if a report was sent.
+// Depending on a caching and reporting configuration the service might decide to skip a report.
+// To make sure report is being sent regardless of circumstances set the force argument to true.
+func (s *service) SendBatteryAlarmReport(force bool) (bool, error) {
```

Comment:
I wonder if providing a report should be a responsibility of a controller. What are the scenarios for an alarm? Is the list finite or they may occur some custom scenarios (so custom alarms would be needed). If there's a fixed number of alarms, maybe the service should be responsible for observing a controller and send an alarm if needed? Let me know what you think.

Reply:
I think we can leave it as it is, however, we should also add some consts. Person implementing should not guess activ/deactiv and low_battery string.

Reply:
Added constants for activ/deactiv/low_battery.